File: src/main/java/com/tms/csp/ast/Imp.java
package com.tms.csp.ast;


import com.tms.csp.fm.dnnf.products.Cube;

public class Imp extends PosComplexMultiVar {

    public static final PosOp OP = PosOp.IMP;

    public Exp or;


    public Imp(Space space, int expId, Exp[] args) {
        super(space, expId, args);
    }

    public boolean isBinaryType() {
        return true;
    }

    @Override
    public Exp condition(Cube ctx) {
        if (isVarDisjoint(ctx)) return this;

        Exp e1 = getExpr1();
        Exp e2 = getExpr2();

        Exp v1 = e1.condition(ctx);
        Exp v2 = e2.condition(ctx);


        if (v1.isFalse() || v2.isTrue()) {
            return mkTrue();
        }


        if (v1.isTrue()) {
            return v2.condition(ctx);
        }

        if (v2.isTrue()) {
            return mkTrue();
        }


        if (v2.isFalse()) {
            return v1.flip().condition(ctx);
        }
        if (v1 == e1 && v2 == e2) {
            return this;
        }

        return mkImp(v1, v2).condition(ctx);
    }


    public Exp toCnf() {

        if (arg2().isAnd()) {
            Exp[] a = new Exp[arg2().size()];
            Exp e1 = arg1();
            for (int i = 0; i < arg2().getArgCount(); i++) {
                Exp e2 = arg2().getArg(i);
                a[i] = mkOr(e1.flip(), e2);
            }
            return mkAnd(a);
        }


//        if (arg1().isVar() && arg2().isCube()) {
//            Exp[] a = new Exp[arg2().argCount() + 1];
//            a[0] = arg1().flip();
//            for (int i = 0; i < arg2().args().length; i++) {
//                a[i + 1] = arg2().getArg(i);
//            }
//            return new Or(a);
//        } else if (arg2().isVar() && arg1().isCube()) {
//            Exp[] a = new Exp[arg1().argCount() + 1];
//            a[0] = arg2();
//            for (int i = 0; i < arg1().args().length; i++) {
//                a[i + 1] = arg1().getArg(i).flip();
//            }
//            return new Or(a);
//        }

        return mkOr(getArg1().flip(), getArg2());
    }

    @Override
    public PosOp getPosOp() {
        return OP;
    }


    public Exp toOr() {
        if (or == null) {

            Exp a1 = getArg1();
            Exp a2 = getArg2();
            Exp na1 = a1.flip();

            Space space = getSpace();

            this.or = space.mkOr(na1, a2);
        }
        return or;
    }


}

File: src/main/java/com/smartsoft/csp/ast/Imp.kt
package com.smartsoft.csp.ast


import com.smartsoft.csp.fm.dnnf.products.Cube

class Imp(space: Space, expId: Int, args: Array<Exp>) : PosComplexMultiVar(space, expId, args) {

    var or: Exp? = null

    override val isBinaryType: Boolean
        get() = true

    override val posOp: PLConstants.PosOp
        get() = OP

    override fun condition(ctx: Cube): Exp {
        if (isVarDisjoint(ctx)) return this

        val e1 = expr1
        val e2 = expr2

        val v1 = e1.condition(ctx)
        val v2 = e2.condition(ctx)


        if (v1.isFalse || v2.isTrue) {
            return mkTrue()
        }


        if (v1.isTrue) {
            return v2.condition(ctx)
        }

        if (v2.isTrue) {
            return mkTrue()
        }


        if (v2.isFalse) {
            return v1.flip.condition(ctx)
        }
        return if (v1 === e1 && v2 === e2) {
            this
        } else mkImp(v1, v2).condition(ctx)

    }


    fun toCnf(): Exp {
        if (arg2.isAnd) {
            val a: Array<Exp?> = arrayOfNulls<Exp>(arg2.size())
            val e1 = arg1
            for (i in 0 until arg2.argCount) {
                val e2 = arg2.getArg(i)
                a[i] = mkOr(e1.flip, e2)
            }
            return mkAnd(*a.requireNoNulls())
        }

        return mkOr(arg1.flip, arg2)
    }


    val toOr: Or
        get() {
            if (or == null) {

                val a1 = arg1
                val a2 = arg2
                val na1 = a1.flip

                val space = space

                this.or = space.mkOr(na1, a2)
            }
            return or!!.asOr
        }

    companion object {

        val OP: PLConstants.PosOp = PLConstants.PosOp.IMP
    }


}
